src/yaml-support/yaml-schema.ts

```diff
@@ -0,0 +1,165 @@
+        return;
+    }
+    // register for kubernetes schema provider
+    yamlPlugin.registerContributor('kubernetes', (resource) => {
+        const textEditor = vscode.window.visibleTextEditors.find((editor) => editor.document.uri.toString() === resource);
+        if (textEditor) {
+            const yamlDocs = yamlLocator.getYamlDocuments(textEditor.document);
+            let choices: string[] = [];
+            yamlDocs.forEach((doc)=> {
+                // if the yaml document contains apiVersion and kind node, it will report it is a kubernetes yaml
+                // file
+                const rootMapping = doc.nodes.find((node) => node.kind === 'MAPPING');
+                // TODO, unwrap quotes
+                const apiVersion = rootMapping.mappings.find((map) => map.key && map.key.raw === 'apiVersion');
+                const kind = rootMapping.mappings.find((map) => map.key && map.key.raw === 'kind');
+                if (apiVersion && kind) {
+                    choices.push(apiVersion.value.raw + GROUP_VERSION_KIND_SEPARATOR + kind.value.raw);
+                }
+            });
+            choices = _.uniq(choices);
+            if (choices.length > 1) {
+                return 'kubernetes://combine/' + _.uniq(choices).join('+');
+            }
+            if (choices.length === 1) {
+                return 'kubernetes://manifest/'  + choices[0];
+            }
+            return undefined;
+        }
+    }, (uri) => {
+        let _uri = Uri.parse(uri);
+        if (!_uri.path || ! _uri.path.startsWith('/')) {
+            return '';
+        }
+        if (_uri.path.includes('+')) {
+            const manifestRef = _uri.path.slice(1).split('+').map((r) => {
+                return {$ref: `kubernetes://manifest/${r}`};
+            });
+            return jsonToString({ oneOf: [...manifestRef] });
+        }
+        let manifestType = _uri.path.slice(1);
+        //apiVersionKind[0] is apiVersion and apiVersionKind[1] is kind
+        const apiVersionKind = manifestType.split(GROUP_VERSION_KIND_SEPARATOR);
+        if (_definitions[manifestType]) {
+            return jsonToString(_definitions[manifestType]);
+        }
+        const matchedSchema = _.values(_definitions).find((schema) => schema.apiVersion && schema.kind && apiVersionKind[1] === schema.kind &&
+                apiVersionKind[0] === schema.apiVersion);
+        if (matchedSchema) {
+            return jsonToString(matchedSchema);
+        }
+        return undefined;
+    });
+}
+
+// convert '#/definitions/com.github.openshift.origin.pkg.build.apis.build.v1.ImageLabel' to
+// 'com.github.openshift.origin.pkg.build.apis.build.v1.ImageLabel'
+const getNameInDefinitions = ($ref: string): string => {
+    const prefix = '#/definitions/';
+    if ($ref.includes(prefix)) {
+        return $ref.slice($ref.indexOf(prefix) + prefix.length);
+    } else {
+        throw new Error('bad reference name.');
```

Comment:
This error is not helpful if the user ever sees it.  (If it's a "can't happen" error then please add a test that it _doesn't_.  **grin**)